### DISCUSSION TITLE
2020/11/30/off-heap-reconnaissance/

### DISCUSSION THREAD

**Discussion by bric3** [bric3/bric3.github.io#42] [Blog Comments]
# 2020/11/30/off-heap-reconnaissance/
Off-Heap memory reconnaissance
https://blog.arkey.fr/2020/11/30/off-heap-reconnaissance/

**Comment by bric3** [comment]
Originally posted by @maulonobile on 2021-09-20T07:19:28Z

----

WOW!!!  
Thank you for this article.

**Comment by bric3** [comment]
Originally posted by Stefan Cordes (RSC) on 2022-03-16T13:20:17Z

----

Thanks for the great post about all the memory details, especially [https://blog.arkey.fr/2020/...](https://blog.arkey.fr/2020/11/30/off-heap-reconnaissance/#_paging_and_the_java_heap "https://blog.arkey.fr/2020/11/30/off-heap-reconnaissance/#_paging_and_the_java_heap")  
With the "free" RSS = "Resident Set Size" kept in physical memory.

**Comment by bric3** [comment]
Originally posted by @xxliu on 2022-06-23T02:06:15Z

----

Nice article! thanks!   
It is worth noting that jdk8u uses mtInternal instead of mtOther for nio DirectByteBuffer.

[https://github.com/corretto...](https://github.com/corretto/corretto-8/blob/develop/hotspot/src/share/vm/prims/unsafe.cpp#L624 "https://github.com/corretto/corretto-8/blob/develop/hotspot/src/share/vm/prims/unsafe.cpp#L624")

**Reply by bric3** [reply]
@xxliu Hi   
Thanks for the feedback !

I believe this is mentioned in this [section](https://blog.arkey.fr/2020/11/30/off-heap-reconnaissance/#_the_jvm_component_memory "https://blog.arkey.fr/2020/11/30/off-heap-reconnaissance/#_the_jvm_component_memory") on point 10, before JDK11 NMT reports Direct Byte Buffers in the Internal section.

**Comment by bric3** [comment]
Originally posted by @jumarko on 2022-07-26T11:12:23Z

----

Excellent, in-depth article - thanks a lot. I wish there were more resources like this.

I have a couple of comments:

1.NMT tracking overhead.  
> Enabling detailed native memory tracking (NMT) causes a 5% to 10% performance overhead. The summary mode merely has an impact in memory usage as shown below and is usually enough.

Is this really true? Do you have an authoritative source of information? In the JDK docs I only found a general statement about the 5-10% performance penalty but not really what it means (CPU utilization, memory overhead, ...) or whether it applies only to _detailed_ tracking mode or also to _summary_.

2. Native vs heap memory.  
You say:   
> The RSS is 4.6 GiB, and it's Java heap size is 4.25 GiB, indicating this process uses around 0.35 GiB of non-Java heap memory, I'll refer to this memory as native memory.

And obviously you expand on this a lot later in the post, including the lazy paging mechanism.  
However, just reading the sentence it is confusing - as you demonstrate later, large portion of the heap (of that 4.25 GiB) might not be resident in RAM. Thus we cannot really tell how much of the RSS (4.6 GiB) is actually heap vs non-heap based on this simple fact and the statement is misleading.  
Also it should be "its Java heap size" not "it's Java heap size" :).

**Reply by bric3** [reply]
Originally posted by @jumarko on 2022-07-30T05:35:24Z

----

Re: NMT tracking overhead.  
I posted a question on stackoverflow and got an excellent answer from Andrei Pangin who I highly respect: [https://stackoverflow.com/a...](https://stackoverflow.com/a/73167790/1184752 "https://stackoverflow.com/a/73167790/1184752")

**Reply by bric3** [reply]
Wow, sorry for taking that long to reply, so yes NMT in `summary` mode is barely noticeable in general, there's a very small memory impact on each allocations, but since the JVM is mostly doing Arena memory management, and as NMT shows there's some malloc calls, but in the end this is quite limited and small in comparison with server app memory (I would even say lambdas). CPU wise I'm not quite sure but I don't think that's a lot and it only happens on allocations, or when querying that info. So it feels like the lower bar, in my experience this looked like less that 5% on today's hardware, on GCP vhost (share with other containers).

For the `details` mode, I'd say it should be in the higher bar, so 10%. Actually I've only played with it to help me understand a few things. Native allocations are not supposed to happen as much as the Java app, but since most of the allocation stacks happen in native code, this could have shown visible impacts as JVM subsystem would have to be paused a tad more in order to capture this stack. And then the stack frame has to be stored, so in the end I assumed it was not a good fit for  
production.

**Reply by bric3** [reply]
Actually just tool a look at the SO answers, great answers by the way by *JDK masters*. I should update this post, with quotes form these. Thanks for linking !

**Reply by bric3** [reply]
About the remark 2, I took an iterative approach to this issue, hence this quite not true statement. I can adjust the phrasing, thanks for the feedback.

**Comment by MahmoudFawzyKhalil** [comment]
Absolutely fantastic article! The equation at the end is super useful to quickly diagnose what memory is going where for when the RSS > NMT we check memory mapped files and memory allocation by native libraries. 

*Thank you!* 😁

**Reply by MahmoudFawzyKhalil** [reply]
I did find the part with pmap that shows 

native heap memory heap
java heap

confusing.

I am not sure what is the difference between "native heap memory heap" and "java heap" in the pmap output?

**Reply by bric3** [reply]
Thank you.

The `pmap` tool will display memory mappings the application requested to the OS (Linux). All these mappings can be considered native as it's the JVM or some of it's libraries that does so.

In these native mappings there is a memory range where the _Java heap_ is "hosted". And of course the JVM manages how Java objects are allocated within this range.

Now the concept of _heap_ also exists in the native world, as the native world programs can allocate memory either on stack or on the _heap_, the later can be done via `malloc` (and cleared via `free`) - or possibly in C++ when creating new objects (but this is not in the specification). In other words what does `malloc is allocating in a _memory mapped_ range known as the _heap_, which I'm refering to as _native heap_.

Hopes that clear that confusion.
